dali/sasha/saarch.cpp

```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+        }
+        void setSDSConnection(Owned<IRemoteConnection> &conn, const char *mask)
+        {
+            if (isWild)
+            {
+                if (!conn)
+                    conn.setown(querySDS().connect(baseXPath.str(), myProcessSession(), 0, 5*60*1000)); // connection to all
+                return;
+            }
+
+            VStringBuffer xpath("%s/%s", baseXPath.str(), mask);
             conn.setown(querySDS().connect(xpath.str(), myProcessSession(), 0, 5*60*1000));
         }
-        if (conn)
+        bool isOnline(Owned<IRemoteConnection> &conn, const char *wuid, const char *mask)
         {
-            Owned<IPropertyTreeIterator> iter = conn->queryRoot()->getElements(isWild ? "*" : NULL);
-            unsigned index = 0;
-            StringBuffer val;
-            ForEach(*iter) {
-                IPropertyTree &pt=iter->query();
-                const char *wuid = pt.queryName();
-                if (index>start+num)
-                    break;
-        //      PROGLOG("match before=%s after=%s wuid=%s",before.str(),after.str(),wuid);
-                if ((!mask||!*mask||!isWild||WildMatch(wuid,mask,true)) &&
-                    ((before.length()==0)||(stricmp(wuid,before)<0)) &&
-                    ((after.length()==0)||(stricmp(wuid,after)>=0))) {
-        //          PROGLOG("matched before=%s after=%s wuid=%s",before.str(),after.str(),wuid);
-                    bool hasowner = owner&&*owner;
-                    bool hascluster = cluster&&*cluster;
-                    bool hasstate = state&&*state;
-                    bool hasjobname = jobname&&*jobname;
-                    bool hasoutput = outputformat&&*outputformat;
-                    bool inrange = (index>=start)&&(index<start+num);
-                    if (hasowner||hasstate||hascluster||hasjobname||(hasoutput&&inrange)) {
-                        try {
-                            if (hasowner&&(!pt.getProp("@submitID",val.clear())||!WildMatch(val.str(),owner,true)))
-                                continue;
-                            if (hasstate&&(!pt.getProp("@state",val.clear())||!WildMatch(val.str(),state,true)))
-                                continue;
-                            if (hascluster&&(!pt.getProp("@clusterName",val.clear())||!WildMatch(val.str(),cluster,true)))
-                                continue;
-                            if (hasjobname&&(!pt.getProp("@jobName",val.clear())||!WildMatch(val.str(),jobname,true)))
+            setSDSConnection(conn, mask);
+            return isWild ? conn->queryRoot()->hasProp(wuid) : conn != nullptr;
+        }
+        bool checkFilters(IPropertyTree *t, bool onlineWU)
+        {
+            //The following code is a clone of the pre-existing code.
+            const char *owner = cmd->queryOwner();
+            const char *cluster = cmd->queryCluster();
+            const char *jobName = cmd->queryJobName();
+            const char *state = cmd->queryState();
```

Comment:
Perhaps these should be members as checkFilters called often and these are also needed/used in other places, e.g. getWUs()

Reply:
Ok, I will add them as members .